### PR TITLE
polishes and final additions to user management

### PR DESCRIPTION
**Pull request by JulianFP**
This is whats new:

- all negative responses made by us now contain an errorTypes attribute. The errorTypes are defined as follows:
  - serverConfig
      - error due to configuration of server 
      - e.g.: disableSignup is set in config
  - email
      - error due to email being invalid in some way
      - e.g. when setting email in signup, changeEmail
      - not for login or similar, only when it is NOT an authentication paramter! (see auth for that)!
  - password
      - error due to password being invalid in some way 
      - e.g. when signup, changePassword (like password too short)
      - not for login or any kind of authentication, only when it is NOT an authentication parameter (see auth for that)!
  - permission
      - user doesn't have permission to access this route/resource/feature
      - e.g. admin exclusive routes, user tries to modify other users or user not activated even though that is required for route
      - often combined with http status code 403
      - not for authentication failures, see auth for that!
  - auth
      - authentication failed for some reason
      - e.g. invalid credentials, invalid password, invalid token, etc.
  - notInDatabase
      - requested ressource is not in database 
      - e.g. when user with specified email couldn't be found in database 
  - operation
      - invalid operation at this time (e.g. some condition for this operation that is not part of request itself is not met)
      - e.g. activate_user when user is already activated
- add three custom function decorators for api routes:
  - activatedRequired: Checks wheither the user is activated before they can use the route. Mainly for upcoming job management routes
  - confirmIdentity: Checks the password from a provided 'password' form attribute (additionally to JWT Token). For routes that have serious consequences like deleteUser, changeUserEmail, changeUserPassword
  - emailModifyForAdmins: If there is a form attribute called 'emailModify': Checks if user is admin and if the specified user exists. If yes, returns user specified with 'emailModify'. If this attribute doesn't exist it will just return the currently logged in user. Useful for routes like changeUserEmail, changeUserPassword that can be used both by admins and non-admins
- Add CORS support: All routes will have the 'Access-Control-Allow-Origin' set to '*' and thus allowing all origins to access it
- add new route resendActivationEmail: resending this mail over changeUserEmail route is awkward, this should have a dedicated route
- add new route invalidateAllTokens: invalidates all existing tokens of the current user thus logging all its sessions out
- sanitize usage of get/post among all routes
  - get: for safe and idempotent routes
  - post: for routes that are neither safe nor idempotent
  - (put: for routes that are not safe but idempotent, not used currently)
- move all user management routes to /api/user subroute to avoid repeated usage of 'user' in route name and to make it coherent with /api/jobs and /api/runners
- tests
  - move all tests of user management into its own file called test_user-management so that tests for the jobs and runner apis can be separated more easily in the future 
  - add new tests for new routes,
  - add new test cases for existing routes (that check for bugs I found)
  - adjust tests for new features/behavior
- fixing various other bugs

This required the following new dependency:
- flask-cors: for CORS support

### CODE REVIEW

**Comment by JulianFP** [comment]
This PR should be ready now, sorry. The 7 additional untested lines codecov complains about are the new wrapper that tests if a user is activated. That is not tested yet because it is not in use yet, it is for the /api/jobs routes that are not in this branch

File: project_W/app.py [hunk 1]
```diff
@@ -63,32 +65,29 @@ def decode_key_loader(_jwt_header, jwt_data):
         # If the user doesn't exist then the token will fail anyways
         return JWT_SECRET_KEY
 
-    @app.post("/api/signup")
+    @app.post("/api/user/signup")
```
Comment (MarkusEverling) [inline project_W/app.py:68]: Seeing as the prefixes for job/runner management are `/api/jobs` and `/api/runners`, should the user management prefix be `/api/users`? Or should we make the other prefixes singular as well?

File: project_W/app.py [hunk 2]
```diff
@@ -99,157 +98,123 @@ def login():
...
 
         logger.info(" -> login successful, returning JWT token")
-        return jsonify(access_token=create_access_token(user))
+        return jsonify(msg="Login successful", access_token=create_access_token(user))
 
-    @app.post("/api/requestPasswordReset")
+    @app.get("/api/user/requestPasswordReset")
     def requestPasswordReset():
-        email = request.form['email']
+        email = request.args.get("email", type=str)
```
Comment (MarkusEverling) [inline project_W/app.py:108]: This will return `None` instead of throwing if the key is missing from the arguments. Is that intentional?

**Review by MarkusEverling — COMMENTED** [review]
Just two small things, LGTM otherwise.

File: project_W/app.py (hunk 1 above)
Comment (JulianFP) [reply project_W/app.py:68]: Yeah I know what you mean. What I had in my mind is that all these routes just apply to one user at a time (for example: /api/user/info returns the userinfo of this one specific user) while the /api/jobs api often describes multiple jobs at a time (like /api/jobs/list which returns the list of all jobs of a user). But you are right, it does seem inconsistent. Should I change it?

File: project_W/app.py (hunk 2 above)
Comment (JulianFP) [reply project_W/app.py:108]: No it is not, I will fix it. BTW, the experience for when you forget to provide some arguments to a post/get request is currently a bit rough in general, the response is anything but helpful. But that is not that important I think, I don't really feel like going through all routes and add if conditions with nice responses right now :D

File: project_W/app.py (hunk 1 above)
Comment (MarkusEverling) [reply project_W/app.py:68]: I don't have a strong opinion on it, just do what seems most logical to you I guess.

**Review by MarkusEverling — APPROVED** [review]

**Pull request merged**